pkg/controllers/nodeclass/instanceprofile_test.go

```diff
@@ -164,4 +181,282 @@ var _ = Describe("NodeClass InstanceProfile Status Controller", func() {
+
+		nodeClass = ExpectExists(ctx, env.Client, nodeClass)
+		newProfileName := nodeClass.Status.InstanceProfile
+
+		// Verify both profiles exist but new one is used
+		Expect(awsEnv.IAMAPI.InstanceProfiles).To(HaveLen(2))
+		Expect(newProfileName).NotTo(Equal(initialProfileName))
+		Expect(awsEnv.IAMAPI.InstanceProfiles[newProfileName].Roles).To(HaveLen(1))
+		Expect(*awsEnv.IAMAPI.InstanceProfiles[newProfileName].Roles[0].RoleName).To(Equal("role-B"))
+	})
+
+	It("should mark new instance profile as protected", func() {
+		// Create NodeClass with role-A
+		nodeClass.Spec.Role = "role-A"
+		ExpectApplied(ctx, env.Client, nodeClass)
+		ExpectObjectReconciled(ctx, env.Client, controller, nodeClass)
+
+		// Get profile name
+		nodeClass = ExpectExists(ctx, env.Client, nodeClass)
+		profileName := nodeClass.Status.InstanceProfile
+
+		// Verify profile exists and is protected
+		Expect(awsEnv.IAMAPI.InstanceProfiles).To(HaveLen(1))
+		Expect(awsEnv.InstanceProfileProvider.IsProtected(profileName)).To(BeTrue())
+	})
+
+	It("should mark old instance profile as protected", func() {
+		// Initial setup with role-A
+		nodeClass.Spec.Role = "role-A"
+		ExpectApplied(ctx, env.Client, nodeClass)
+		ExpectObjectReconciled(ctx, env.Client, controller, nodeClass)
+
+		// Get initial profile name
+		nodeClass = ExpectExists(ctx, env.Client, nodeClass)
+		oldProfileName := nodeClass.Status.InstanceProfile
+
+		// Remove the created profile from protectedProfiles cache to ensure
+		// that it is again added once we switch to "role-B"
+		awsEnv.InstanceProfileProvider.SetProtectedState(oldProfileName, false)
+		Expect(awsEnv.InstanceProfileProvider.IsProtected(oldProfileName)).To(BeFalse())
+
+		// Change to role-B
+		nodeClass.Spec.Role = "role-B"
+		ExpectApplied(ctx, env.Client, nodeClass)
+		ExpectObjectReconciled(ctx, env.Client, controller, nodeClass)
+
+		nodeClass = ExpectExists(ctx, env.Client, nodeClass)
+		newProfilename := nodeClass.Status.InstanceProfile
+
+		// Verify old profile is still protected as well as new one
+		Expect(awsEnv.InstanceProfileProvider.IsProtected(oldProfileName)).To(BeTrue())
+		Expect(awsEnv.InstanceProfileProvider.IsProtected(newProfilename)).To(BeTrue())
+	})
+
+	It("should use the cached instance profile when spec.role is changed", func() {
+		ExpectApplied(ctx, env.Client, nodeClass)
+
+		// Set up cache with a profile for role-A
+		cachedProfileName := "cached-profile"
+		awsEnv.RecreationCache.SetDefault(fmt.Sprintf("%s/%s", "role-A", nodeClass.UID), cachedProfileName)
+
+		// Apply NodeClass with role-A
+		nodeClass.Spec.Role = "role-A"
+		ExpectApplied(ctx, env.Client, nodeClass)
+		ExpectObjectReconciled(ctx, env.Client, controller, nodeClass)
+
+		nodeClass = ExpectExists(ctx, env.Client, nodeClass)
+
+		// Verify it used the cached profile
+		Expect(nodeClass.Status.InstanceProfile).To(Equal(cachedProfileName))
+		Expect(awsEnv.IAMAPI.InstanceProfiles).To(HaveLen(0))
```

Comment:
I was going to call this out as problematic, but I think this get's fixed with the suggestion to always verify that the instance profile exists and that the role is correct. We don't want to get into a state where we use a non-existent role just because it's in the cache.